src/main/java/net/dv8tion/jda/internal/JDAImpl.java

```diff
@@ -480,22 +485,56 @@ public long getGatewayPing()
     public JDA awaitStatus(@Nonnull Status status, @Nonnull Status... failOn) throws InterruptedException
     {
         Checks.notNull(status, "Status");
-        Checks.check(status.isInit(), "Cannot await the status %s as it is not part of the login cycle!", status);
         if (getStatus() == Status.CONNECTED)
             return this;
-        List<Status> failStatus = Arrays.asList(failOn);
-        while (!getStatus().isInit()                         // JDA might disconnect while starting
-                || getStatus().ordinal() < status.ordinal()) // Wait until status is bypassed
+
+        MiscUtil.tryLock(statusLock);
+        try
         {
-            if (getStatus() == Status.SHUTDOWN)
-                throw new IllegalStateException("Was shutdown trying to await status.\nReason: " + shutdownReason);
-            else if (failStatus.contains(getStatus()))
-                return this;
-            Thread.sleep(50);
+            EnumSet<Status> endCondition = EnumSet.of(status, failOn);
+            Status current = getStatus();
+            while (!current.isInit()                      // In case of disconnects during startup
+                 || current.ordinal() < status.ordinal()) // If we missed the status (e.g. LOGGING_IN -> CONNECTED happened while waiting for lock)
+            {
+                if (current == Status.SHUTDOWN)
+                    throw new IllegalStateException("Was shutdown trying to await status.\nReason: " + shutdownReason);
+                if (endCondition.contains(current))
+                    return this;
+
+                statusCondition.await();
+                current = getStatus();
+            }
         }
+        finally
+        {
+            statusLock.unlock();
+        }
+
         return this;
     }
 
+    @Override
+    public boolean awaitShutdown(@Nonnull Duration timeout) throws InterruptedException
+    {
+        long deadline = timeout.isZero() || timeout.isNegative() ? Long.MAX_VALUE : System.currentTimeMillis() + timeout.toMillis();
+        MiscUtil.tryLock(statusLock);
+        try
+        {
+            Status current = getStatus();
+            while (current != Status.SHUTDOWN)
+            {
+                if (!statusCondition.await(deadline - System.currentTimeMillis(), TimeUnit.MILLISECONDS))
```

Comment:
why does this compute a deadline as timeout + currentTimeMillis just to remove it later?
Couldn't we just do this?:
```suggestion
        long deadline = timeout.isZero() || timeout.isNegative() ? Long.MAX_VALUE : timeout.toMillis();
        MiscUtil.tryLock(statusLock);
        try
        {
            Status current = getStatus();
            while (current != Status.SHUTDOWN)
            {
                if (!statusCondition.await(deadline, TimeUnit.MILLISECONDS))
```

Reply:
The condition can wake multiple times and each time it would wait the same timeout, which would result in a total timeout that is longer than the provided one.

Reply:
Yup, makes sense